File: srcs/django_code/front/js/game/TournamentMenu.js
import { canvas, ctx } from "./canvas.js";
import { Button } from "./Button.js";
import { Wall } from "./Wall.js";
import { Player } from "./Player.js";
import { Obstacle } from "./Obstacle.js";
import { Ball } from "./Ball.js";
import { AI } from "./AI.js";
import { Vec2 } from "./Vec2.js";
import { is_colliding } from "./Hitbox.js";
import { Tournament } from "./Tournament.js";

export class TournamentMenu {
	constructor() {
		this.size = [canvas.width * 0.2, canvas.height * 0.1];
		this.mod_size = [canvas.width * 0.11, canvas.height * 0.07];
		this.score = 10;
		this.ai_nb = 0;
		this.max_ai = 2;
		this.nb_players = 2;
		this.down_buttons = [new Button("BACK TO MENU", this.size[0] * 0.1, canvas.height - (this.size[1] * 1.4), this.size[0], this.size[1]),
				new Button("START", canvas.width - (this.size[0] * 1.1), canvas.height - (this.size[1] * 1.4), this.size[0], this.size[1])];
		this.mod_buttons = [new Button("LOCAL", canvas.width / 5 - (this.mod_size[0] / 2), canvas.height / 3 - (this.mod_size[1] / 2), this.mod_size[0], this.mod_size[1]),
              	new Button("ONLINE", canvas.width / 5 * 2 - (this.mod_size[0] / 2), canvas.height / 3 - (this.mod_size[1] / 2), this.mod_size[0], this.mod_size[1]),
          		new Button("BORDERLESS", canvas.width / 5 * 3 - (this.mod_size[0] / 2), canvas.height / 3 - (this.mod_size[1] / 2), this.mod_size[0], this.mod_size[1]),
            	new Button("OBSTACLE", canvas.width / 5 * 4 - (this.mod_size[0] / 2), canvas.height / 3 - (this.mod_size[1] / 2), this.mod_size[0], this.mod_size[1])];
		this.param_buttons = [new Button("-", canvas.width * 0.335, canvas.height / 3 * 2 - (canvas.height * 0.025), canvas.width * 0.015, canvas.height * 0.03),
				new Button("+", canvas.width * 0.355, canvas.height / 3 * 2 - (canvas.height * 0.025), canvas.width * 0.015, canvas.height * 0.03),
				new Button("-", canvas.width * 0.56, canvas.height / 3 * 2 - (canvas.height * 0.025), canvas.width * 0.015, canvas.height * 0.03),
				new Button("+", canvas.width * 0.58, canvas.height / 3 * 2 - (canvas.height * 0.025), canvas.width * 0.015, canvas.height * 0.03),
				new Button("-", canvas.width * 0.835, canvas.height / 3 * 2 - (canvas.height * 0.025), canvas.width * 0.015, canvas.height * 0.03),
				new Button("+", canvas.width * 0.855, canvas.height / 3 * 2 - (canvas.height * 0.025), canvas.width * 0.015, canvas.height * 0.03)];
		this.mod_buttons[0].highlight = true;
	}

	draw() {
		ctx.fillText("TOURNAMENT", canvas.width / 2, canvas.height * 0.1);
		for (let b of this.down_buttons)
			b.draw();
		ctx.font = Math.floor(canvas.height * 0.06) + "px pong-teko";
		ctx.fillText("MATCH SCORE = " + this.score, canvas.width / 4, canvas.height / 3 * 2);
		ctx.fillText("PLAYERS = " + this.nb_players, canvas.width / 2, canvas.height / 3 * 2);
		ctx.fillText("AI OPPONENTS = " + this.ai_nb, canvas.width / 4 * 3, canvas.height / 3 * 2);
		for (let b of this.mod_buttons)
			b.draw();
		for (let b of this.param_buttons)
			b.draw();
		ctx.font = Math.floor(canvas.height * 0.085) + "px pong-teko";
	}

	click(core, mousePos) {
		const pos = new Vec2(mousePos[0], mousePos[1]);
		for (let b of this.down_buttons) {
			if (is_colliding(pos, [0, 0], b.hitbox.pos, this.size)) {
				if (b.name === "BACK TO MENU") {
					core.state = "menu";
					core.mode = "none";
					core.max_score = 10;
					core.tournament_menu = false;
					break;
				}
				else if (b.name === "START") {
					this.start(core);
					break;
				}
			}
		}
		for (let b of this.mod_buttons) {
			if (is_colliding(pos, [0, 0], b.hitbox.pos, this.mod_size)) {
				if (b.name === "LOCAL" && !b.highlight)
					this.mod_buttons[1].highlight = false;
				else if (b.name === "ONLINE" && !b.highlight)
					this.mod_buttons[0].highlight = false;
				b.highlight = !b.highlight;
				break;
			}
		}
		for (let b of this.param_buttons) {
			if (is_colliding(pos, [0, 0], b.hitbox.pos, [b.width, b.height])) {
				if (b === this.param_buttons[0] || b === this.param_buttons[1]) {
					if (b.name === "-" && this.score > 1)
						this.score -= 1;
					else if (b.name === "+")
						this.score += 1;
				}
				else if (b === this.param_buttons[2] || b === this.param_buttons[3]) {
					if (b.name === "-" && this.nb_players > 2) {
						this.nb_players -= 1;
						if (this.ai_nb > this.nb_players)
							this.ai_nb = this.nb_players;
					}
					else if (b.name === "+")
						this.nb_players += 1;
					this.max_ai = this.nb_players;
				}
				else {
					if (b.name === "-" && this.ai_nb > 0)
						this.ai_nb -= 1;
					else if (b.name === "+" && this.ai_nb < this.max_ai)
						this.ai_nb += 1;
				}
				break;
			}
		}
		if (this.ai_nb >= this.nb_players - 1) {
			this.mod_buttons[1].highlight = false;
			this.mod_buttons[0].highlight = true;
		}
	}

	start(core) {
		if (!this.validStart())
			return;
		this.getMods(core.alias);
		core.max_score = this.score;
		if (this.mod_list.includes("LOCAL")) {
			this.initPlayers(core);
			this.initWalls(core);
			core.ball = new Ball((this.mod_list.includes("BORDERLESS")) ? true : false);
			core.state = "tournament";
			core.mode = "LOCAL";
			if (this.mod_list.includes("OBSTACLE"))
				core.obstacle = new Obstacle();
			const msg = {"type" : 'tournament', 'online' : 'false'};
			core.GameHub.send(JSON.stringify(msg));
		}
		else if (this.mod_list.includes("ONLINE")) {
			const msg = {"type" : 'tournament', 'online' : 'true', 'mods' : this.mod_list, 'score' : this.score, 'ai' : this.ai_nb, 'players' : Object.keys(this.players).length};
			core.GameHub.send(JSON.stringify(msg));
			core.online = true;
		}

		core.tournament = new Tournament(this.mod_list, this.nb_players, this.ai_nb, this.score, core.online);
	}

	getMods(alias = "Player") {
		this.mod_list = [];
		for (let b of this.mod_buttons) {
			if (b.highlight)
				this.mod_list.push(b.name);
		}
		this.players = {};
		for (let i = 0; i < this.nb_players; i++) {
			this.players[i + 1] = (this.ai_nb < this.nb_players - i) ? alias + (i + 1).toString() : "AI";
		}
	}

	initPlayers(core) {
		core.players = [];
		for (const key in this.players)
			core.players.push(new Player(parseInt(key), this.players[key], Object.keys(this.players).length, this.mod_list.includes("BORDERLESS"), false));
		for (let p of core.players)
			if (p.name === "AI")
				core.ai.push(new AI(p));
	}

	initWalls(core) {
		if (this.mod_list.includes("BORDERLESS"))
			core.walls = false;
		else
			core.walls = [new Wall("up", false), new Wall("down", false)];
	}

	validStart() {
		for (let b of this.mod_buttons) {
			if (b.highlight && b.name === "LOCAL") {
				return true;
			}
			if (b.highlight && b.name === "ONLINE") {
				return true;
			}
		}
		return false;
	}

	responsive() {
		this.size = [canvas.width * 0.2, canvas.height * 0.1];
		this.mod_size = [canvas.width * 0.11, canvas.height * 0.07];

		this.down_buttons = [new Button("BACK TO MENU", this.size[0] * 0.1, canvas.height - (this.size[1] * 1.4), this.size[0], this.size[1]),
				new Button("START", canvas.width - (this.size[0] * 1.1), canvas.height - (this.size[1] * 1.4), this.size[0], this.size[1])];
		this.mod_buttons = [new Button("LOCAL", canvas.width / 5 - (this.mod_size[0] / 2), canvas.height / 3 - (this.mod_size[1] / 2), this.mod_size[0], this.mod_size[1]),
              	new Button("ONLINE", canvas.width / 5 * 2 - (this.mod_size[0] / 2), canvas.height / 3 - (this.mod_size[1] / 2), this.mod_size[0], this.mod_size[1]),
          		new Button("BORDERLESS", canvas.width / 5 * 3 - (this.mod_size[0] / 2), canvas.height / 3 - (this.mod_size[1] / 2), this.mod_size[0], this.mod_size[1]),
            	new Button("OBSTACLE", canvas.width / 5 * 4 - (this.mod_size[0] / 2), canvas.height / 3 - (this.mod_size[1] / 2), this.mod_size[0], this.mod_size[1])];
		this.param_buttons = [new Button("-", canvas.width * 0.335, canvas.height / 3 * 2 - (canvas.height * 0.025), canvas.width * 0.015, canvas.height * 0.03),
				new Button("+", canvas.width * 0.355, canvas.height / 3 * 2 - (canvas.height * 0.025), canvas.width * 0.015, canvas.height * 0.03),
				new Button("-", canvas.width * 0.56, canvas.height / 3 * 2 - (canvas.height * 0.025), canvas.width * 0.015, canvas.height * 0.03),
				new Button("+", canvas.width * 0.58, canvas.height / 3 * 2 - (canvas.height * 0.025), canvas.width * 0.015, canvas.height * 0.03),
				new Button("-", canvas.width * 0.835, canvas.height / 3 * 2 - (canvas.height * 0.025), canvas.width * 0.015, canvas.height * 0.03),
				new Button("+", canvas.width * 0.855, canvas.height / 3 * 2 - (canvas.height * 0.025), canvas.width * 0.015, canvas.height * 0.03)];
		this.mod_buttons[0].highlight = true;
	}
}

File: Game Client/game/TournamentMenu.py
from .config import *
from .Button import *
from .Wall import *
from .Player import *
from .Obstacle import *
from .Ball import *
from .Vec2 import *
from .StartScreen import *
from .TournamentNames import *
from .Tournament import *

class TournamentMenu:
	def __init__(self):
		self.size = [winWidth * 0.2, winHeight * 0.1]
		self.mod_size = [winWidth * 0.11, winHeight * 0.07]
		self.score = 10
		self.ai_nb = 0
		self.max_ai = 2
		self.nb_players = 2
		self.down_buttons = [Button("BACK TO MENU", self.size[0] * 0.1, winHeight - (self.size[1] * 1.4), self.size[0], self.size[1], winHeight * 0.085),
				Button("START", winWidth - (self.size[0] * 1.1), winHeight - (self.size[1] * 1.4), self.size[0], self.size[1], winHeight * 0.085)]
		self.mod_buttons = [Button("LOCAL", winWidth / 5 - (self.mod_size[0] / 2), winHeight / 3 - (self.mod_size[1] / 2), self.mod_size[0], self.mod_size[1], winHeight * 0.06),
              	Button("ONLINE", winWidth / 5 * 2 - (self.mod_size[0] / 2), winHeight / 3 - (self.mod_size[1] / 2), self.mod_size[0], self.mod_size[1], winHeight * 0.06),
          		Button("BORDERLESS", winWidth / 5 * 3 - (self.mod_size[0] / 2), winHeight / 3 - (self.mod_size[1] / 2), self.mod_size[0], self.mod_size[1], winHeight * 0.06),
            	Button("OBSTACLE", winWidth / 5 * 4 - (self.mod_size[0] / 2), winHeight / 3 - (self.mod_size[1] / 2), self.mod_size[0], self.mod_size[1], winHeight * 0.06)]
		self.param_buttons = [Button("-", winWidth * 0.335, winHeight / 3 * 2 - (winHeight * 0.025), winWidth * 0.015, winHeight * 0.03, winHeight * 0.06),
				Button("+", winWidth * 0.355, winHeight / 3 * 2 - (winHeight * 0.025), winWidth * 0.015, winHeight * 0.03, winHeight * 0.06),
				Button("-", winWidth * 0.56, winHeight / 3 * 2 - (winHeight * 0.025), winWidth * 0.015, winHeight * 0.03, winHeight * 0.06),
				Button("+", winWidth * 0.58, winHeight / 3 * 2 - (winHeight * 0.025), winWidth * 0.015, winHeight * 0.03, winHeight * 0.06),
				Button("-", winWidth * 0.785, winHeight / 3 * 2 - (winHeight * 0.025), winWidth * 0.015, winHeight * 0.03, winHeight * 0.06),
				Button("+", winWidth * 0.805, winHeight / 3 * 2 - (winHeight * 0.025), winWidth * 0.015, winHeight * 0.03, winHeight * 0.06)]
		self.mod_buttons[0].highlight = True
  
	def draw(self, win):
		text_font = pg.font.Font(font, int(winHeight * 0.085))
		text = text_font.render("TOURNAMENT", True, (255, 255, 255))
		win.blit(text, [winWidth / 2 - (text.get_size()[0] / 2), winHeight * 0.1 - (text.get_size()[1] / 2)])
		for b in self.down_buttons:
			b.draw(win)
		text_font = pg.font.Font(font, int(winHeight * 0.06))
		text = text_font.render("MATCH SCORE = " + str(self.score), True, (255, 255, 255))
		win.blit(text, [winWidth / 4 - (text.get_size()[0] / 2), (winHeight / 3 * 2) - (text.get_size()[1] / 2)])
		text = text_font.render("PLAYERS = " + str(self.nb_players), True, (255, 255, 255))
		win.blit(text, [winWidth / 2 - (text.get_size()[0] / 2), (winHeight / 3 * 2) - (text.get_size()[1] / 2)])
		text = text_font.render("AI = " + str(self.ai_nb), True, (255, 255, 255))
		win.blit(text, [(winWidth / 4 * 3) - (text.get_size()[0] / 2), (winHeight / 3 * 2) - (text.get_size()[1] / 2)])
		for b in self.mod_buttons:
			b.draw(win)
		for b in self.param_buttons:
			b.draw(win)

	async def click(self, core, mousePos):
		for b in self.down_buttons:
			if b.hitbox.collidepoint(mousePos):
				if b.name == "BACK TO MENU":
					core.state = "menu"
					core.mode = "none"
					core.max_score = 10
					core.tournament_menu = False
				elif b.name == "START":
					await self.start(core)
				break

		for b in self.mod_buttons:
			if b.hitbox.collidepoint(mousePos):
				if (b.name == "LOCAL" and not b.highlight):
					self.mod_buttons[1].highlight = False
				elif (b.name == "ONLINE" and not b.highlight):
					self.mod_buttons[0].highlight = False
				b.highlight = not b.highlight
				break

		for b in self.param_buttons:
			if b.hitbox.collidepoint(mousePos):
				if b == self.param_buttons[0] or b == self.param_buttons[1]:
					if b.name == '-' and self.score > 1:
						self.score -= 1
					elif b.name == '+':
						self.score += 1
				elif b == self.param_buttons[2] or b == self.param_buttons[3]:
					if b.name == '-' and self.nb_players > 2:
						self.nb_players -= 1
						if self.ai_nb > self.nb_players:
							self.ai_nb = self.nb_players
					elif b.name == '+':
						self.nb_players += 1
					self.max_ai = self.nb_players
				else:
					if b.name == '-' and self.ai_nb > 0:
						self.ai_nb -= 1
					elif b.name == '+' and self.ai_nb < self.max_ai:
						self.ai_nb += 1
				break

		if self.ai_nb >= self.nb_players - 1:
			self.mod_buttons[1].highlight = False
			self.mod_buttons[0].highlight = True
   
	async def start(self, core):
		if not self.validStart():
			return
		self.getMods(core.alias)
		core.max_score = self.score
		core.tournament = Tournament(self.mod_list, self.nb_players, self.ai_nb, self.score, core.online, Player(1, core.alias, 2, "BORDERLESS" in self.mod_list, False))
		core.tournament_id = 1
		if "LOCAL" in self.mod_list:
			self.initPlayers(core)
			self.initWalls(core)
			core.ball = Ball("BORDERLESS" in self.mod_list)
			core.state = "tournament names"
			core.mode = "LOCAL"
			if "OBSTACLE" in self.mod_list:
				core.obstacle = Obstacle()
			msg = {"type" : 'tournament', 'online' : 'false'}
			core.GameHub.send(json.dumps(msg))
			core.tournament_names = TournamentNames(core.players)
		elif "ONLINE" in self.mod_list:
			msg = {"type" : 'tournament', 'online' : 'true', 'mods' : self.mod_list, 'score' : self.score, 'ai' : self.ai_nb, 'players' : self.players.__len__()}
			core.players = [Player(1, core.alias, 2, "BORDERLESS" in self.mod_list, False)]
			core.GameHub.send(json.dumps(msg))
			response : dict = json.loads(await core.GameHub.recv())
			core.GamePort = response['port']
			core.mode = "ONLINE"
			core.state = "tournament"
			core.id = response['pos']
			core.tournament_id = core.id
			core.online = True
			core.tournament.id = response['ID']

		core.start_screen = StartScreen(core.mode, core.online)

	def getMods(self, alias = "Player"):
		self.mod_list = []
		for b in self.mod_buttons:
			if b.highlight:
				self.mod_list.append(b.name)
		self.players = {}
		for i in range(self.nb_players):
			self.players[i + 1] = alias + str(i + 1) if self.ai_nb < self.nb_players - i else "AI"
   
	def initPlayers(self, core):
		core.players = []
		for key, name in self.players.items():
			core.players.append(Player(key, name, self.players.__len__(), "BORDERLESS" in self.mod_list, False))

	def initWalls(self, core):
		if ("BORDERLESS" in self.mod_list):
			core.walls = False
		else:
			core.walls = [Wall("up", False), Wall("down", False)]

	def validStart(self):
		for b in self.mod_buttons:
			if b.highlight and (b.name == "LOCAl" or b.name == "ONLINE"):
				return True
		return False